src/ngraph_encapsulate_op.cc

```diff
@@ -309,6 +309,60 @@ class NGraphEncapsulateOp : public OpKernel {
 
       auto function_size = ng_function->get_graph_size() / 1024;  // kb unit
 
+      // Get the serialized ops and stored the allreduce ops to a vector and
+      ng::NodeVector allreduce_op_list;
+      for (const shared_ptr<ng::Node>& node : ng_function->get_ordered_ops()) {
+        if (node->get_name().compare(0, 9, "AllReduce") == 0) {
+          allreduce_op_list.push_back(node);
+          // Diagnosis to check the input, dependencies, and output for each
+          // Allreduce op
+          if (std::getenv("NGRAPH_DISABLE_LOGGING") == nullptr) {
+#if defined NGRAPH_DISTRIBUTED
```

Comment:
This "#if defined NGRAPH_DISTRIBUTED" should be moved above allreduce_op_list is initialized?

Reply:
Builder would be a better place to add the control dependencies in the graph. You can move the code here. 
https://github.com/NervanaSystems/ngraph-tf/blob/master/src/ngraph_builder.cc#L4328

Once the function is created you can add the control dependencies.